static/src/javascripts/projects/common/modules/spacefinder.ts

```diff
@@ -0,0 +1,448 @@
+// total_hours_spent_maintaining_this = 81.5
+
+import { memoize } from 'lodash-es';
+import { isInVariantSynchronous } from 'common/modules/experiments/ab';
+import { spacefinderOkrMegaTest } from 'common/modules/experiments/tests/spacefinder-okr-mega-test';
+import { noop } from 'lib/noop';
+import fastdom from '../../../lib/fastdom-promise';
+import { mediator } from '../../../lib/mediator';
+import { markCandidates } from './mark-candidates';
+import { onImagesLoadedBroken } from './on-images-loaded-broken.js';
+import { onImagesLoadedFixed } from './on-images-loaded-fixed.js';
+
+type RuleSpacing = {
+	minAbove: number;
+	minBelow: number;
+};
+
+type SpacefinderItem = {
+	top: number;
+	bottom: number;
+	element: HTMLElement;
+	meta?: {
+		tooClose: unknown[];
+	};
+};
+
+type SpacefinderRules = {
+	bodySelector: string;
+	body?: HTMLElement | Document;
+	slotSelector: string;
+	// minimum from slot to top of page
+	absoluteMinAbove?: number;
+	// minimum from para to top of article
+	minAbove: number;
+	// minimum from (top of) para to bottom of article
+	minBelow: number;
+	// vertical px to clear the content meta element (byline etc) by. 0 to ignore
+	// used for carrot ads
+	clearContentMeta?: number;
+	// custom rules using selectors.
+	selectors?: Record<string, RuleSpacing>;
+	// will run each slot through this fn to check if it must be counted in
+	filter?: (x: SpacefinderItem, lastWinner?: SpacefinderItem) => boolean;
+	// will remove slots before this one
+	startAt?: HTMLElement;
+	// will remove slots from this one on
+	stopAt?: HTMLElement;
+	// will reverse the order of slots (this is useful for lazy loaded content)
+	fromBottom?: boolean;
+};
+
+type SpacefinderWriter = (paras: HTMLElement[]) => Promise<void>;
+
+type SpacefinderOptions = {
+	waitForLinks?: boolean;
+	waitForImages?: boolean;
+	waitForInteractives?: boolean;
+	debug?: boolean;
+};
+
+type ExcludedItem = SpacefinderItem | HTMLElement;
+
+type SpacefinderExclusions = Record<string, ExcludedItem[]>;
+
+type ElementDimensionMap = Record<string, SpacefinderItem[]>;
+
+type Measurements = {
+	bodyTop: number;
+	bodyHeight: number;
+	candidates: SpacefinderItem[];
+	contentMeta?: SpacefinderItem;
+	opponents?: ElementDimensionMap;
+};
+
+const query = (selector: string, context?: HTMLElement | Document) => [
+	...(context ?? document).querySelectorAll<HTMLElement>(selector),
+];
+
+// maximum time (in ms) to wait for images to be loaded and rich links
+// to be upgraded
+const LOADING_TIMEOUT = 5000;
+
+const defaultOptions = {
+	waitForImages: true,
+	waitForLinks: true,
+	waitForInteractives: false,
+	debug: false,
+};
+
+const isIframe = (node: Node): node is HTMLIFrameElement =>
+	node instanceof HTMLIFrameElement;
+
+const isIframeLoaded = (iframe: HTMLIFrameElement) => {
+	// TODO do we need this try / catch?
+	try {
+		return iframe.contentWindow?.document.readyState === 'complete';
+	} catch (err) {
+		return true;
+	}
+};
```

Comment:
Add a sentinel in the catch block?

Reply:
Are you thinking we should add an `amIUsed` in the catch block to see if an error is ever thrown? And we'd remove the try / catch if we find it isn't.